### PR TITLE
Account management rework

### PR DESCRIPTION
**Pull request by anonymous123-code**
- `accounts.helix.json` now uses a json object which contains the account list and the index of the selected instance. *this will break old versions of the file*
- The corresponding struct also stores the path, so it can be saved using a method
- refactored the account management CLI and reworked it to use the new system

### CODE REVIEW

**Review by Oliver-makes-code — APPROVED** [review]
lgtm

**Review by kb-1000 — CHANGES_REQUESTED** [review]
Please rebase rather than merging the main branch in.

**Comment by anonymous123-code** [comment]
> Please rebase rather than merging the main branch in.

Should I do that for the current PRs or is that for the future?

File: helixlauncher-cli/src/main.rs [hunk 1]
```diff
@@ -219,62 +211,32 @@ fn add_account_callback(code: String, uri: String, message: String) {
...
-                    .as_path(),
-            );
-            if add_acc_res.is_err() {
-                no_print = true;
-            }
-        }
-        if !no_print {
-            println!("Welcome! You are logged in as: {}", username);
-        }
+        .await?;
```
[the remark this answers is not in the document]
Comment (anonymous123-code) [reply helixlauncher-cli/src/main.rs:219]: Anyhow errors generally look pretty good:
grafik

File: helixlauncher-cli/src/main.rs [hunk 2]
```diff
@@ -1,10 +1,10 @@
 //! Helix Launcher CLI
 //! This is an example implementation of the Helix Launcher CLI.
 
-use anyhow::Result;
+use anyhow::{Ok, Result};
```
[the remark this answers is not in the document]
Comment (anonymous123-code) [reply helixlauncher-cli/src/main.rs:4]: No it isnt, I saw it in autocomplete and just assumed it is.

File: helixlauncher-core/src/auth/account.rs [hunk 3]
```diff
@@ -15,34 +15,41 @@ pub enum AccountManagerError {
...
-    Ok(serde_json::from_reader(BufReader::new(
-        match File::open(account_json) {
-            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(vec![]),
-            result => result,
-        }?,
-    ))?)
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct AccountConfig {
+    pub accounts: Vec<Account>,
+    pub selected: usize,
```
[the remark this answers is not in the document]
Comment (anonymous123-code) [reply helixlauncher-core/src/auth/account.rs:29]: Hmm, should I rename it to `selected_uuid`, too?

[the remark this answers is not in the document]
Comment (kb-1000) [reply helixlauncher-core/src/auth/account.rs:29]: Well, if the UUID is the "id" of the account we use, then it would make sense not to, too

**Comment by kb-1000** [comment]
> Should I edit the commit history for the current PRs or is that for the future?

I'll just squash merge it when it's ready, you don't need to

File: helixlauncher-core/src/auth/account.rs [hunk 4]
```diff
@@ -15,34 +15,42 @@ pub enum AccountManagerError {
...
+    /// The UUID of the selected account
+    pub selected: String,
+    #[serde(skip)]
+    path: PathBuf,
 }
 
-pub fn add_account(mut account: Account, account_json: &Path) -> Result<(), AccountManagerError> {
-    let mut accounts: Vec<Account> = get_accounts(account_json)?;
-    if accounts.is_empty() {
-        account.selected = true
```
Comment (kb-1000) [inline helixlauncher-core/src/auth/account.rs:40]: This feature appears to be missing in the new implementation: defaulting to the single account in the list if there is only one (though that should probably be done in the place it's used in, if the selected default is `None`)

**Review by kb-1000 — CHANGES_REQUESTED** [review]

**Review by kb-1000 — APPROVED** [review]
There's still a few nitpicks but I can change those post-merge.
Otherwise, LGTM.

**Comment by anonymous123-code** [comment]
I now fixed the error when saving a deserialized, and added a corresponding test case.

**Pull request merged**